spec/mem/tree.md

```diff
@@ -0,0 +1,562 @@
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn node_access(
+        &mut self, 
+        curr: BorTag,
+        access_pattern: AccessPattern,
+        ptr: Pointer<TreeBorrowsProvenance>,
+        size: Size,
+        base: BorTag,
+        node_relation: NodeRelation
+    ) -> Result {
+        let Some(mut node) = self.tree.nodes.get(curr) else { throw_ub!("Tree Borrows: node not existed"); };
+
+        // Update the permission for each location
+        for location in ptr.addr..ptr.addr + size.bytes() {
+            let Some(curr_permission) = node.permissions.get(location) else {
+                throw_ub!("Tree Borrows: location without permission");
+            };
+
+            let next_permission = Self::permission_transition(curr_permission, access_pattern, node_relation)?;
+            node.permissions.insert(location, next_permission);
+        }
+
+        // Insert back the updated node
+        self.tree.nodes.insert(curr, node);
+
+        // Recursively updating descendants of the `curr`
+        let new_relation = if curr == base { NodeRelation::Child } else { node_relation };
+        for child_tag in node.children {
+            self.node_access(child_tag, access_pattern, ptr, size, base, new_relation)?;
+        }
+        ret(())
+    }
+
+    /// Recusively do permission transition starting from the root node.
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn tree_access(
+        &mut self, 
+        base: BorTag, 
+        access_pattern: AccessPattern, 
+        ptr: Pointer<TreeBorrowsProvenance>,
+        size: Size
+    ) -> Result {
+        let root = self.tree.root_tag;
+        let node_relation = if root == base { NodeRelation::Child } else { NodeRelation::Foreign };
+        self.node_access(self.tree.root_tag, access_pattern, ptr, size, base, node_relation)?;
+        ret(())
+    }
+}
+```
+
+Here we define some helper methods to implement the memory interface.
+
+```rust
+impl<T: Target> TreeBorrowsMemory<T> {
+    /// Given the permission, the start location and the size of location, create an initialized permission map.
+    fn init_range_permissions(permission: Permission, location_start: Location, alloc_size: Size) -> Map<Location, Permission> {
+        let mut child_permissions = Map::new();
+        for location in location_start..location_start+alloc_size.bytes() {
+            child_permissions.insert(location, permission);
+        }
+        child_permissions
+    }
+
+    /// Create a new node for a pointer (reborrow)
+    fn new_child(
+        &mut self, 
+        ptr: Pointer<TreeBorrowsProvenance>,
+        alloc_size: Size, 
+        permission: Permission
+    ) -> Result<Pointer<TreeBorrowsProvenance>> {
+        let Some((parent_tag, alloc_id)) = ptr.provenance else {
+            throw_ub!("Tree Borrows: Pointer does not contain provenance");
```

Comment:
If the pointee is zero, then this should not be UB.

Reply:
This is fixed by the above
```
// For zero-sized accesses, there is nothing to do.
if pointee_size.is_zero() {
    return ret(ptr);
}
```